### PR TITLE
[web] Add ShaderBuilder, change drawVertices to use builder.

### PR DESCRIPTION
**Pull request by ferhatb**
## Description

Moved hard coded vertex/fragment shader code into a builder.
Prepares code for future sweepgradient.

## Related Issues

https://github.com/flutter/flutter/issues/41389

## Tests

shape_builder_test.dart

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Reviewer Checklist

- [x] I have submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

### CODE REVIEW

File: lib/web_ui/lib/src/engine/html/render_vertices.dart [hunk 1]
```diff
@@ -161,16 +116,18 @@ class _WebGlRenderer implements _GlRenderer {
     if (widthInPixels == 0 || heightInPixels == 0) {
       return;
     }
+    String vertexShader = _writeVerticesVertexShader();
```
Comment (nturgut) [inline lib/web_ui/lib/src/engine/html/render_vertices.dart:119]: nit: Looks like these 2 strings can be final.

File: lib/web_ui/lib/src/engine/html/render_vertices.dart [hunk 2]
```diff
@@ -161,16 +116,18 @@ class _WebGlRenderer implements _GlRenderer {
...
+    String fragmentShader = _writeVerticesFragmentShader();
+    print(vertexShader);
```
Comment (nturgut) [inline lib/web_ui/lib/src/engine/html/render_vertices.dart:121]: suggestion: Are these from debugging? If not we can maybe print them as 'INFO: <a message>:<value>'

File: lib/web_ui/lib/src/engine/html/render_vertices.dart [hunk 3]
```diff
@@ -186,18 +143,20 @@ class _WebGlRenderer implements _GlRenderer {
     assert(positionsBuffer != null); // ignore: unnecessary_null_comparison
     gl.bindArrayBuffer(positionsBuffer);
     gl.bufferData(positions, gl.kStaticDraw);
+    Object? positionLoc = gl.getAttribLocation(glProgram.program, 'position');
     js_util.callMethod(
-        gl.glContext!, 'vertexAttribPointer', <dynamic>[0, 2, gl.kFloat, false, 0, 0]);
+        gl.glContext!, 'vertexAttribPointer', <dynamic>[
```
Comment (nturgut) [inline lib/web_ui/lib/src/engine/html/render_vertices.dart:148]: nit: if we add a trailing comma the formatting will be:
```
gl.glContext!, 'vertexAttribPointer', <dynamic>[
    positionLoc, 2, gl.kFloat, false, 0, 0,
]);
```
(I'm told it's also in the style guide :) )

File: lib/web_ui/lib/src/engine/html/render_vertices.dart (hunk 1 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/render_vertices.dart:119]: Done.

File: lib/web_ui/lib/src/engine/html/render_vertices.dart (hunk 2 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/render_vertices.dart:121]: Yes debugging. Removed.

File: lib/web_ui/lib/src/engine/html/render_vertices.dart (hunk 3 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/render_vertices.dart:148]: Done.

File: lib/web_ui/lib/src/engine/html/render_vertices.dart [hunk 4]
```diff
@@ -209,6 +168,51 @@ class _WebGlRenderer implements _GlRenderer {
     context.restore();
   }
 
+  // Vertex shader transforms pixel space [Vertices.positions] to
```
Comment (nturgut) [inline lib/web_ui/lib/src/engine/html/render_vertices.dart:170]: suggestion: here in and in the next method we can use doc comments (///)

File: lib/web_ui/test/engine/surface/shaders/shader_builder_test.dart [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+      '1.2, 2.2, 3.2, // second column\n'
+      '1.3, 2.3, 3.3  // third column\n'
+      ')';
+  String mat4Sample = 'mat3(1.1, 2.1, 3.1, 4.1,\n'
+      '1.2, 2.2, 3.2, 4.2,\n'
+      '1.3, 2.3, 3.3, 4.3,\n'
+      '1.4, 2.4, 3.4, 4.4,\n'
+      ')';
+
+  setUpAll(() async {
```
Comment (nturgut) [inline lib/web_ui/test/engine/surface/shaders/shader_builder_test.dart:27]: I think this won't work on ios-safari due to simulator restrictions.

We can add a skip to this test methods using operating system, example: https://github.com/flutter/engine/blob/bada9fc5f34b0145e7a1a38719a48971f23e7019/lib/web_ui/test/engine/surface/surface_test.dart#L205

**Review by nturgut — APPROVED** [review]
LGTM.

Do you think it would be useful if we also add a golden test for chrome? Or is there already one that covers this case? (we can also do it with the other changes for sweepgradient)

File: lib/web_ui/test/engine/surface/shaders/shader_builder_test.dart (hunk 5 above)
Comment (ferhatb) [reply lib/web_ui/test/engine/surface/shaders/shader_builder_test.dart:27]: See https://github.com/flutter/goldens/blob/master/engine/web/draw_vertices_triangle_fan.iOS_Safari.png . webgl1 seems to be working in that case.

**Comment by ferhatb** [comment]
> LGTM.
> 
> Do you think it would be useful if we also add a golden test for chrome? Or is there already one that covers this case? (we can also do it with the other changes for sweepgradient)

Thanks for checking. draw_vertices_test is the golden one.

File: lib/web_ui/lib/src/engine/html/render_vertices.dart (hunk 4 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/render_vertices.dart]: Done.

File: lib/web_ui/test/engine/surface/shaders/shader_builder_test.dart (hunk 5 above)
Comment (nturgut) [reply lib/web_ui/test/engine/surface/shaders/shader_builder_test.dart:27]: I see so they fixed the bug on the simulator :) this is good to know let me try to see if I can un-skip some other tests :D

**Pull request merged**